### PR TITLE
E2EE: implement receiving of the messages

### PR DESCRIPTION
**Pull request by a-andreyev**
With the current state, it is possible to receive the (megolm) messages at the room.
Created sessions are not saving to persistent storage though (the crypto storage for the sessions is not implemented yet), so to check it's working I'm logging in as new device with clean new cache and olm account every time.
Olm decryption is also implemented to initialize the megolm session to share messages.
Tested with Quaternion. Apropriate draft commit to display encrypted messages: https://github.com/quotient-im/Quaternion/commit/9f9868ace6d72c1ec53fbe9f69737247e5fdacca

### CODE REVIEW

**Comment by a-andreyev** [comment]
Small refresh: rebased with the master branch and also implemented `Quotient_ENABLE_E2EE` CMake option to make building E2EE optional. (Contributes to #369)

**Comment by a-andreyev** [comment]
Small refresh: rebased with the master branch and also implemented `Quotient_ENABLE_E2EE` CMake option to make building E2EE optional. (Contributes to #369)

File: CMakeLists.txt [hunk 1]
```diff
@@ -55,22 +57,27 @@ endif()
...
 get_filename_component(Qt5_Prefix "${Qt5_DIR}/../../../.." ABSOLUTE)
 
-if ((NOT DEFINED USE_INTREE_LIBQOLM OR USE_INTREE_LIBQOLM)
-        AND EXISTS ${PROJECT_SOURCE_DIR}/3rdparty/libQtOlm/lib/utils.h)
-    add_subdirectory(3rdparty/libQtOlm EXCLUDE_FROM_ALL)
-    include_directories(3rdparty/libQtOlm)
-    if (NOT DEFINED USE_INTREE_LIBQOLM)
-        set (USE_INTREE_LIBQOLM 1)
+if (${PROJECT_NAME}_ENABLE_E2EE)
+    add_definitions(-DE2EE_ENABLED)
```
Comment (KitsuneRal) [inline CMakeLists.txt]: I'd prefer to rather see [`target_compile_definitions`](https://cmake.org/cmake/help/v3.10/command/target_compile_definitions.html#command:target_compile_definitions) here, something like
```suggestion
    target_compile_definitions(${PROJECT_NAME} PUBLIC E2EE_ENABLED)
```
Don't have an opinion on whether the preprocessor symbol should be namespaced, like `QUOTIENT_E2EE_ENABLED`.

File: CMakeLists.txt [hunk 2]
```diff
@@ -108,18 +115,22 @@ if (ABS_API_DEF_PATH AND ABS_GTAD_PATH)
...
+                OUTPUT_VARIABLE QTOLM_GIT_SHA1
+                OUTPUT_STRIP_TRAILING_WHITESPACE)
+            message( STATUS "  Library git SHA1: ${QTOLM_GIT_SHA1}")
+        endif (GIT_FOUND)
+    else ()
+        message( STATUS "Using libQtOlm ${QtOlm_VERSION} at ${QtOlm_DIR}")
+    endif ()
 else ()
-    message( STATUS "Using libQtOlm ${QtOlm_VERSION} at ${QtOlm_DIR}")
+    message( STATUS "Not going to search for libQtOlm")
```
Comment (KitsuneRal) [inline CMakeLists.txt]: Just drop this branch, rather.

File: lib/connection.cpp [hunk 3]
```diff
@@ -1070,7 +1093,12 @@ QByteArray Connection::accessToken() const { return d->data->accessToken(); }
 
 QtOlm::Account* Connection::olmAccount() const
```
Comment (KitsuneRal) [inline lib/connection.cpp]: This should not even be in the API if E2EE is off, maybe? (because how do we know wth `QtOlm::Account` is)

Comment (a-andreyev) [reply lib/connection.cpp]: Agree. It's interesting that it looks like I'm not using proposed API anywhere yet %) Should be double-checked why do I proposed it later.

File: CMakeLists.txt (hunk 1 above)
Comment (a-andreyev) [reply CMakeLists.txt]: Not sure either, decided to stick with `QUOTIENT_E2EE_ENABLED`

**Comment by a-andreyev** [comment]
I agree, thank you! It great you found a time to take a look. Fixed the issues and rebased to save commits' history.

File: CMakeLists.txt [hunk 4]
```diff
@@ -159,6 +167,7 @@ set(lib_SRCS
     lib/events/directchatevent.cpp
     lib/events/encryptionevent.cpp
     lib/events/encryptedevent.cpp
+    lib/events/roomkeyevent.cpp
```
Comment (KitsuneRal) [inline CMakeLists.txt]: Should also come to `libquotient.pri`?

Comment (a-andreyev) [reply CMakeLists.txt]: Thank you! You right, added header and source path to `.pri`

**Pull request merged**